crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -78,21 +78,52 @@ impl Fulfillment {
             Some(FeePolicy::Surplus {
                 factor,
                 max_volume_factor,
+            }) => self.calculate_fee(
+                self.order().sell.amount.0,
+                self.order().buy.amount.0,
+                prices,
+                *factor,
+                *max_volume_factor,
+            ),
+            Some(FeePolicy::PriceImprovement {
+                factor,
+                max_volume_factor,
+                quote,
             }) => {
-                let fee_from_surplus = self.fee_from_surplus(prices, *factor)?;
-                let fee_from_volume = self.fee_from_volume(prices, *max_volume_factor)?;
-                // take the smaller of the two
-                tracing::debug!(uid=?self.order().uid, fee_from_surplus=?fee_from_surplus, fee_from_volume=?fee_from_volume, protocol_fee=?(std::cmp::min(fee_from_surplus, fee_from_volume)), executed=?self.executed(), surplus_fee=?self.surplus_fee(), "calculated protocol fee");
-                Ok(std::cmp::min(fee_from_surplus, fee_from_volume))
+                let sell_amount = quote.sell_amount.max(self.order().sell.amount.0);
+                let buy_amount = quote.buy_amount.min(self.order().buy.amount.0);
+                self.calculate_fee(sell_amount, buy_amount, prices, *factor, *max_volume_factor)
             }
             Some(FeePolicy::Volume { factor }) => self.fee_from_volume(prices, *factor),
             None => Ok(0.into()),
         }
     }
 
-    fn fee_from_surplus(&self, prices: ClearingPrices, factor: f64) -> Result<eth::U256, Error> {
-        let sell_amount = self.order().sell.amount.0;
-        let buy_amount = self.order().buy.amount.0;
+    /// Computes protocol fee compared to the given reference amounts taken from
+    /// the order or a quote.
+    fn calculate_fee(
+        &self,
+        reference_sell_amount: eth::U256,
+        reference_buy_amount: eth::U256,
+        prices: ClearingPrices,
+        factor: f64,
+        max_volume_factor: f64,
+    ) -> Result<eth::U256, Error> {
+        let fee_from_surplus =
+            self.fee_from_surplus(reference_sell_amount, reference_buy_amount, prices, factor)?;
+        let fee_from_volume = self.fee_from_volume(prices, max_volume_factor)?;
+        // take the smaller of the two
+        tracing::debug!(uid=?self.order().uid, fee_from_surplus=?fee_from_surplus, fee_from_volume=?fee_from_volume, protocol_fee=?(std::cmp::min(fee_from_surplus, fee_from_volume)), executed=?self.executed(), surplus_fee=?self.surplus_fee(), "calculated protocol fee");
+        Ok(std::cmp::min(fee_from_surplus, fee_from_volume))
+    }
+
+    fn fee_from_surplus(
```

Comment:
I believe so far we overlooked the case where theoretically multiple partial fills with a fee than is less than the volume cap can in total exceed the volume cap.
For example if we have 2 partial fills that take 0.5% of the total volume as a fee every additional partial fill should not incur any fees since it would exceed the 1% in total.
Should probably not be addressed in this PR but we should probably solve this eventually. Although this admittedly sounds pretty contrived.

Reply:
> Although this admittedly sounds pretty contrived.

That doesn't sound contrived to me :)
Looks like the most obvious solution is to implement a cumulative fee tracking mechanism. @sunce86, should we create an issue for this?

Reply:
`fee_from_volume` is calculated as a percentage of the executed (partial) amount and not as a percentage of the total order volume so I don't think this can happen.